### PR TITLE
Make it possible to configure WSSE like you do in SoapUI

### PR DESCRIPTION
**Pull request by veewee**
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | yes
| Fixed issues | #3, #4

#### Todo

- [x] Tests
- [x] Docs
- [x] Examples

#### Awaiting

- https://github.com/robrichards/wse-php/pull/67

#### Summary

Current implementation is not very flexible and a lot of edge cases are not covered.
This new approach makes it possible to configure WSSE [similar to like you do in  soap-ui](https://www.soapui.org/docs/soapui-projects/ws-security/):

```php
<?php

use Soap\Psr18WsseMiddleware\WSSecurity\KeyStore\Key;
use Soap\Psr18WsseMiddleware\WSSecurity\SignatureAlgorithm;
use Soap\Psr18WsseMiddleware\WSSecurity\SignatureDigestAlgorithm;
use Soap\Psr18WsseMiddleware\WSSecurity\KeyIdentifier;
use Soap\Psr18WsseMiddleware\WsseMiddleware;
use Soap\Psr18WsseMiddleware\WSSecurity\Entry;
use VeeWee\Xml\Dom\Document;

$x509 = Certificate::fromFile('cert.pem')
    ->withPassphrase('xxx');
$wsseRequestToken = new WsseMiddleware(
    outgoing: [
        new Entry\Timestamp(60),
        new Entry\BinarySecurityToken($pubKey),
        (new Entry\Signature(
            $privKey,
            new KeyIdentifier\BinarySecurityTokenIdentifier()
        ))
            ->withSignatureMethod(SignatureMethod::RSA_SHA256)
            ->withDigestMethod(DigestMethod::SHA256)
            ->withSignAllHeaders(true)
            ->withSignBody(true)
    ]
);

$wsseAuthorizedRequest = new WsseMiddleware(
    outgoing: [
        new Entry\Timestamp(300),
        (new Entry\Signature(
            $x509,
            new KeyIdentifier\SamlKeyIdentifier($samlAssertionId)
        ))
            ->withSignatureAlgorithm(SignatureAlgorithm::RSA_SHA256)
            ->withDigestAlgorithm(SignatureDigestAlgorithm::SHA256)
            ->withSignAllHeaders(true)
            ->withSignBody(true),
        new Entry\SamlAssertion($saml),
    ]
);

```

### CODE REVIEW

File: src/SamlMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
+<?php
```
Comment (veewee) [inline src/SamlMiddleware.php:1]: TODO : just as example - to remove

File: src/WSSecurity/Entry/Signature.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+final class Signature implements WsseEntry
+{
+    private KeyInterface $privateKey;
+    private KeyIdentifier $keyIdentifier;
+
+    private SignatureAlgorithm $signatureAlgorithm = SignatureAlgorithm::RSA_SHA1;
+    private SignatureDigestAlgorithm $digestAlgorithm = SignatureDigestAlgorithm::SHA1;
+
+    private $signAllHeaders = false;
```
Comment (veewee) [inline src/WSSecurity/Entry/Signature.php:23]: todo : better defaults and toggling flags on signing specific headers

File: src/WSSecurity/KeyIdentifier/BinarySecurityToken.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+declare(strict_types=1);
+
+namespace Soap\Psr18WsseMiddleware\WSSecurity\KeyIdentifier;
+
+use DOMElement;
+use RobRichards\WsePhp\WSSESoap;
+use Soap\Psr18WsseMiddleware\WSSecurity\KeyStore\KeyInterface;
+use VeeWee\Xml\Dom\Document;
+
+class BinarySecurityToken implements KeyIdentifier
```
Comment (veewee) [inline src/WSSecurity/KeyIdentifier/BinarySecurityToken.php:11]: Should adding the binary token go in a separate entry? In soap-ui it is more like a selection of the key identifier.
Not sure if this is reusable for both encruption and signing : to investigate

File: src/WsseMiddleware.php [hunk 4]
```diff
@@ -7,79 +7,48 @@
...
+    private bool $mustUnderstand = false;
+    private ?string $actor = null;
+
+    /**
+     * @no-named-arguments
+     * @param list<WsseEntry> $outgoing
+     * @param list<WsseEntry> $incoming
+     */
+    public function __construct(
+        array $outgoing = [],
```
Comment (veewee) [inline src/WsseMiddleware.php:33]: Alternatively, an OutgoingEntry or IncomingEntry interface could be wrapped arround a WsseEntry.
Not sure what the best way is yet.

By explicitely specifying it in the constructor, you have full control and there is no magic involved.
However, it requires you to know more about WSSE. Maybe that's not a bad thing...

File: src/WSSecurity/Entry/Encryption.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $key = new XMLSecurityKey($this->symmetricEncodingAlgorithm->value);
+        $key->generateSessionKey();
+
+        $siteKey = new XMLSecurityKey($this->keyEncryptionAlgorithm->value, ['type' => 'public']);
+        $siteKey->passphrase = $this->key->passphrase();
+        $siteKey->loadKey($this->key->contents(), false, $this->key->isCertificate());
+
+        $wsse->encryptSoapDoc($siteKey, $key, [
+            'KeyInfo' => [
```
Comment (veewee) [inline src/WSSecurity/Entry/Encryption.php:36]: similar logic is available for signing. Maybe better to move it out to a separate key identifier instead;

File: src/WSSecurity/SignatureCanonicalization.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?php
+declare(strict_types=1);
+
+namespace Soap\Psr18WsseMiddleware\WSSecurity;
+
+/**
+ * Currently, there is no option to configure this in robrichards/wsse.
+ * It is hardcoded to C14N
```
Comment (veewee) [inline src/WSSecurity/SignatureCanonicalization.php:8]: see https://github.com/robrichards/wse-php/blob/c9611f554d88cbd58935d7da901d84093068168c/src/WSSESoap.php#L221-L227

Comment (veewee) [reply src/WSSecurity/SignatureCanonicalization.php:8]: out of scope for now
